pkg/envoy/ads/stream.go

```diff
@@ -141,42 +148,72 @@ func (s *Server) StreamAggregatedResources(server xds_discovery.AggregatedDiscov
-			// and not requesting when the ackVersion is <= what we last sent.
 			// It is possible however for a proxy to have a version that is higher
 			// than what we last sent. (Perhaps the control plane restarted.)
 			// In that case we want to make sure that we send new responses with
 			// VersionInfo incremented starting with the version which the proxy last had.
-			if ackVersion > proxy.GetLastSentVersion(typeURL) {
-				proxy.SetLastSentVersion(typeURL, ackVersion)
+			if requestVersion > proxy.GetLastSentVersion(typeURL) {
+				proxy.SetLastSentVersion(typeURL, requestVersion)
+				goto update
 			}
 
-			lastNonce := proxy.GetLastSentNonce(typeURL)
-			if lastNonce != "" && discoveryRequest.ResponseNonce == lastNonce {
-				log.Debug().Msgf("Nothing changed for Envoy on Pod with UID=%s since Nonce=%s", proxy.GetPodUID(), discoveryRequest.ResponseNonce)
+			// Compare Nonces
+			// As per protocol, we can ignore any request on the TypeURL stream that has not caught up with last sent nonce, if the
+			// nonce is non-empty.
+			// A newly connecting envoy might send a version and nonce from an old control plane, this is currently mitigated by the
+			// forced update @line 69
+			// TODO: Fix reconnection. Forced update causes other issues, we should stick to the protocol and wait for the first request
+			// and gracefully handle reconnecting envoys.
+			lastNonceSent = proxy.GetLastSentNonce(typeURL)
+			if requestNonce != lastNonceSent {
+				log.Debug().Msgf("Ignoring older nonce: envoy %s TypeURL: %s for nonce: %s, current nonce: %s",
+					proxy.GetCertificateSerialNumber(), typeURL.Short(), requestNonce, lastNonceSent)
 				continue
 			}
 
-			if discoveryRequest.ResponseNonce != "" {
-				log.Debug().Msgf("Received discovery request with Nonce=%s from Envoy on Pod with UID=%s; matches=%t; proxy last Nonce=%s",
-					discoveryRequest.ResponseNonce, proxy.GetPodUID(), discoveryRequest.ResponseNonce == lastNonce, lastNonce)
+			// ----
+			// At this point, there is no error and nonces match, it is guaranteed an ACK with last version.
+			// What's left is to check if the resources listed are the same. If they are not, we must respond
+			// with the new reosurces requested
+			// ----
+
+			// If what we last sent is a superset of what the
+			// requests resources subscribes to, it's ACK and nothing needs to be done.
+			// Otherwise, envoy might be asking us for additional resources that have to be sent along.
+			{
+				resourcesLastSent := proxy.GetLastResourcesSent(typeURL)
+				resourcesRequested := mapset.NewSet()
+				for idx := range discoveryRequest.ResourceNames {
+					resourcesRequested.Add(discoveryRequest.ResourceNames[idx])
+				}
+
+				// Difference returns elemenets of <requested> that are not part of elements of <last sent>
+				requestedResourcesDifference := resourcesRequested.Difference(resourcesLastSent)
+				if requestedResourcesDifference.Cardinality() == 0 {
```

Comment:
This could go into a  function that is called something like `hasDelta()` etc...